### PR TITLE
AVR-8: Implement Power Management

### PR DESCRIPTION
**Pull request by nandojve**
### Contribution description

The Xmega have a Power Management implementation. This moves that code to AVR-8 common and enable Power Management for all variants.

### Testing procedure

This was tested using **examples/thread_duel** and **tests/periph/pm** examples on **atxmega-a1u-xpro** and **atmega328p-xplained-mini** boards. It may requires more tests and any help is welcome.

### Issues/PRs references

This is a preparation for future scheduling and IRQ optimizations.

### CODE REVIEW

**Comment by maribu** [comment]
The order in which app+board pairs are build is pretty random and the CI run stops once the first failure is hit. This could indeed result in pretty random failures, if more than on app fails.

The lasted failure (different modules used when building with `TEST_KCONFIG=1` and without) could be fixed e.g. using

```diff
diff --git a/cpu/avr8_common/Kconfig b/cpu/avr8_common/Kconfig
index 03a0f40b5a..d82c187042 100644
--- a/cpu/avr8_common/Kconfig
+++ b/cpu/avr8_common/Kconfig
@@ -12,6 +12,8 @@ config CPU_ARCH_AVR8
     select HAS_ARCH_AVR8
     select HAS_PERIPH_PM
 
+    select MODULE_AVR8_COMMON_PERIPH if TEST_KCONFIG
+    select MODULE_PM_LAYERED if TEST_KCONFIG
     select MODULE_MALLOC_THREAD_SAFE if TEST_KCONFIG
     select MODULE_TINY_STRERROR_AS_STRERROR if TEST_KCONFIG
     # static C++ constructors need guards for thread safe initialization
@@ -41,6 +43,13 @@ config MODULE_AVR8_COMMON
     help
         AVR-8 common code.
 
+config MODULE_AVR8_COMMON_PERIPH
+    bool
+    depends on TEST_KCONFIG
+    default y
+    help
+        Common peripheral drivers used across different AVR-8 MCU families.
+
 # the atmel port uses stdio_uart by default
 choice STDIO_IMPLEMENTATION
     default MODULE_STDIO_UART
```

This one:

```
-- running on worker riotbuild-numa-2 thread 1, build number 186110.
make: Entering directory '/tmp/dwq.0.2615164153767623/2950a57bddbed179a6e917a6f6caf29d/examples/suit_update'
# Reset package to checkout state.
rm -rf /tmp/dwq.0.2615164153767623/2950a57bddbed179a6e917a6f6caf29d/build/pkg/c25519
rm -rf /tmp/dwq.0.2615164153767623/2950a57bddbed179a6e917a6f6caf29d/build/pkg-build/libcose
rm -rf /tmp/dwq.0.2615164153767623/2950a57bddbed179a6e917a6f6caf29d/build/pkg-build/nanocbor
make: Leaving directory '/tmp/dwq.0.2615164153767623/2950a57bddbed179a6e917a6f6caf29d/examples/suit_update'
make: Entering directory '/tmp/dwq.0.2615164153767623/2950a57bddbed179a6e917a6f6caf29d/examples/suit_update'
make: *** No rule to make target '/tmp/dwq.0.2615164153767623/2950a57bddbed179a6e917a6f6caf29d/examples/suit_update/bin/native/suit_update/fw.1688580754.bin', needed by 'link'.  Stop.
make: *** Waiting for unfinished jobs....
Building application "suit_update" for "native" with MCU "native".
```

is indeed a race condition in the build system that is triggered when compiling that app for `native` with `-j` :/ I can (occasionally) reproduce that one locally. I was hoping that @kaspar030 would take a look :innocent:

**Comment by nandojve** [comment]
This seems ready to go!

**Comment by nandojve** [comment]
I was wondering if there is anything pending on this : )

**Comment by nandojve** [comment]
ping

**Comment by nandojve** [comment]
Ping @MrKevinWeiss : )

File: cpu/atmega_common/periph/i2c.c [hunk 1]
```diff
@@ -226,6 +227,7 @@ void i2c_acquire(i2c_t dev)
 {
     assert(dev < I2C_NUMOF);
 
+    pm_block(4); /* Require clkIO */
```
Comment (MrKevinWeiss) [inline cpu/atmega_common/periph/i2c.c:230]: I feel like these should be protected by the mutex for example:

task 1 - block pm
task 1 - acquire i2c
task 2 - block pm
task 2 - waiting to acquire i2c
task 1 - release i2c
task 1 - unblock pm
task 2 - acquire i2c **(but pm is not blocked)**

**Review by MrKevinWeiss — COMMENTED** [review]
After a quick look things seem OK, maybe some logic issues for your consideration. I will try to test tomorrow when I am in the office.  (I need to also familiarize myself with the whole PM stuff).

File: cpu/atmega_common/periph/i2c.c (hunk 1 above)
Comment (benpicco) [reply cpu/atmega_common/periph/i2c.c:230]: In that case pm_blockers would be set to 2 when both tasks do the `pm_block()`.
When task 1 unblocks pm, pm_blockers is still at 1, so pm is blocked when task 2 acquires the mutex.

Comment (MrKevinWeiss) [reply cpu/atmega_common/periph/i2c.c:230]: Ahh it counts... Good good. Thanks for the clarification.

**Comment by MrKevinWeiss** [comment]
This will need a rebase to get in sync with the github merge queue changes.  @benpicco any blocker?

**Comment by benpicco** [comment]
Nothing from my side - how did the current draw change?

**Comment by MrKevinWeiss** [comment]
Once we get the rebase we can merge (it needs a check-commit-success workflow included)

**Comment by nandojve** [comment]
> Once we get the rebase we can merge (it needs a check-commit-success workflow included)

I simple made a rebase, let's see.

**Comment by nandojve** [comment]
> I am happy so ACK!

Thank you so much @MrKevinWeiss  for invest time to test this.

**Comment by nandojve** [comment]
Ops, it seems there is a new cpu in tree atmega8.

**Review by MrKevinWeiss — COMMENTED** [review]
I don't think we need the config anymore...

```diff
diff --git a/cpu/atxmega/Makefile.features b/cpu/atxmega/Makefile.features
index d566d9ce38..7395748241 100644
--- a/cpu/atxmega/Makefile.features
+++ b/cpu/atxmega/Makefile.features
@@ -12,9 +12,3 @@ FEATURES_PROVIDED += periph_cpuid
 FEATURES_PROVIDED += periph_gpio periph_gpio_irq
 FEATURES_PROVIDED += periph_nvm
 FEATURES_PROVIDED += periph_timer periph_timer_periodic
-
-# Add atxmega configurations. This configuration enables modules that are only available when
-# using Kconfig module modelling
-ifeq (1, $(TEST_KCONFIG))
-  KCONFIG_CPU_CONFIG += $(RIOTCPU)/atxmega/atxmega.config
-endif
```

**Comment by maribu** [comment]
I have a fast machine. I can run the `./dist/tools/insufficient_memory/add_insufficient_memory_board.sh` for all the AVR boards on it. I'll push directly to this branch.

**Comment by MrKevinWeiss** [comment]
@maribu if you don't mind, just amend my commit I think so we don't have so much clutter.

**Comment by maribu** [comment]
I'll also do the rebase to resolve the merge conflict

**Comment by maribu** [comment]
Looks like https://github.com/RIOT-OS/RIOT/pull/20130 contains a bug :-/

**Comment by nandojve** [comment]
Thank you for all help @maribu and @MrKevinWeiss

**Pull request merged**